src/hotspot/os/windows/os_windows.cpp

```diff
@@ -3234,18 +3235,32 @@ static size_t large_page_init_decide_size() {
     return 0;
   }
 
-#if defined(IA32) || defined(AMD64)
-  if (size > 4*M || LargePageSizeInBytes > 4*M) {
-    WARN("JVM cannot use large pages bigger than 4mb.");
-    return 0;
-  }
+  if (!os::win32::is_windows_11_or_greater() && !os::win32::is_windows_server_2022_or_greater()) {
+#if defined(IA32)
+    if (size > 4 * M || LargePageSizeInBytes > 4 * M) {
+      WARN("JVM cannot use large pages bigger than 4mb.");
+      return 0;
+    }
+#elif defined(AMD64)
+    if (!EnableAllLargePageSizes) {
+      if (size > 4 * M || LargePageSizeInBytes > 4 * M) {
+        WARN("JVM cannot use large pages bigger than 4mb.");
+        return 0;
+      }
+    } else {
+      WARN("EnableAllLargePageSizes flag is ignored on Windows versions prior to Windows 11/Windows Server 2022 due to limited support.");
+    }
 #endif
+  }
 
   if (LargePageSizeInBytes > 0 && LargePageSizeInBytes % size == 0) {
     size = LargePageSizeInBytes;
+  } else {
+    WARN1("The JVM cannot use large pages due to either the large page setting not being configured or the specified page size not being a multiple of the minimum large page size (%d), defaulting to minimum page size.", size);
```

Comment:
I think LargePageSizeInBytes has a default value of 0. If that is the case, won't this warning be displayed every time the user doesn't set LargePageSizeInBytes? This would be false because we already have size == GetLargePageMinimum() at this point.

Reply:
When the `LargePageSizeInBytes` JVM parameter is not specified, the warning indicating incorrect large page configuration is expected. This warning is also justified if `LargePageSizeInBytes` is deliberately set to a zero value (`0g/m/k`), as it suggests an intentional decision not to utilize large pages. Conversely, in any other situation, a non-zero value should be assigned to this parameter. It's hard to envision a scenario where setting `LargePageSizeInBytes` to a non-zero value would not supersede a zero value, as the purpose of this flag is to define the large page size for the JVM. Ensuring that `LargePageSizeInBytes` is set to a non-zero value is crucial for the correct configuration of large pages and to avoid any related warnings or errors.

Reply:
Here are two scenarios:

1. **Using `UseLargePages` without specifying `LargePageSizeInBytes`:**

   ```bash
   /cygdrive/c/cygwin64/home/singhnitin/microsoft/openjdk-jdk/build/windows-x86_64-server-slowdebug/jdk/bin/java \
   -XX:+UnlockExperimentalVMOptions \
   -XX:+EnableAllLargePageSizes \
   -XX:+UseLargePages \
   -Xlog:pagesize=debug,gc*,gc+heap=debug,gc+age=trace,safepoint:file=./gc_log_with_large_pages_1g1.log \
   ../../HelloWorld.java
   ```

   Output:
   ```
   OpenJDK 64-Bit Server VM warning: The JVM cannot use large pages due to either the large page setting not being configured or the specified page size not being a multiple of the minimum large page size (2097152), defaulting to minimum page size.
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 251658240
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000404800000 bytes: 17104371712
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 33554432
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 268435456
   Hello, World!
   ```

2. **Passing `LargePageSizeInBytes=0g`:**

   ```bash
   /cygdrive/c/cygwin64/home/singhnitin/microsoft/openjdk-jdk/build/windows-x86_64-server-slowdebug/jdk/bin/java \
   -XX:+UnlockExperimentalVMOptions \
   -XX:+EnableAllLargePageSizes \
   -XX:+UseLargePages \
   -XX:LargePageSizeInBytes=0g \
   -XX:ReservedCodeCacheSize=1g \
   -Xlog:pagesize=debug,gc*,gc+heap=debug,gc+age=trace,safepoint:file=./gc_log_with_large_pages_1g1.log \
   ../../HelloWorld.java
   ```

   Output:
   ```
   OpenJDK 64-Bit Server VM warning: The JVM cannot use large pages due to either the large page setting not being configured or the specified page size not being a multiple of the minimum large page size (2097152), defaulting to minimum page size.
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 251658240
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000404800000 bytes: 17104371712
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 33554432
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 33554432
   OpenJDK 64-Bit Server VM warning: Failed to reserve and commit memory using large pages. req_addr: 0x0000000000000000 bytes: 268435456
   Error: Could not find or load main class .cygdrive.c.cygwin64.home.singhnitin.microsoft.openjdk-jdk.build.windows-x86_64-server-slowdebug.jdk.bin.cygdrive.c.cygwin64.home.singhnitin.microsoft.openjdk-jdk.build.windows-x86_64-server-slowdebug.jdk.bin-XX:ReservedCodeCacheSize=1g
   Caused by: java.lang.ClassNotFoundException: /cygdrive/c/cygwin64/home/singhnitin/microsoft/openjdk-jdk/build/windows-x86_64-server-slowdebug/jdk/bin/cygdrive/c/cygwin64/home/singhnitin/microsoft/openjdk-jdk/build/windows-x86_64-server-slowdebug/jdk/bin-XX:ReservedCodeCacheSize=1g
   ```

Reply:
The warning says "defaulting to minimum page size" but doesn't the implementation actually default to the minimum large page size?

Reply:
Is it also possible to separate the warning message into 2 separate warnings by detecting each case separately? I think that would better explain the behavior of the JVM.

Reply:
> The warning says "defaulting to minimum page size" but doesn't the implementation actually default to the minimum large page size?

This is defaulting to minimum page size which is 4k. Since the specified largepagesize is zero or not being set, minimum page size is utilized.

Reply:
> Is it also possible to separate the warning message into 2 separate warnings by detecting each case separately? I think that would better explain the behavior of the JVM.

This looks right to me.

Reply:
> Is it also possible to separate the warning message into 2 separate warnings by detecting each case separately? I think that would better explain the behavior of the JVM.

@dhanalla  @macarte any comments ?